[clang] Fix array types comparison in getCommonSugaredType

spavloff opened the pull request:
Const-qualification of an array caused by constexpr specifier can produce QualType, where the const qualifier is set both as fast qualifier and as a qualifier of the array element type. It can result in a compiler crash, because such QualType does not compare equal to the same type but without extra qualification.

To avoid the crash, the redundant qualifiers are removed while searching for common sugar.

It fixes https://github.com/llvm/llvm-project/issues/97005 (Clang crashed in ASTContext::getCommonSugaredType).

spavloff committed:
[clang] Fix array types comparison in getCommonSugaredType

Const-qualification of an array caused by constexpr specifier can
produce QualType, where the const qualifier is set both as fast
qualifier and as a qualifier of the array element type. It can result in
a compiler crash, because such QualType does not compare equal to the
same type but without extra qualification.

To avoid the crash, the redundant qualifiers are removed while searching
for common sugar.

It fixes https://github.com/llvm/llvm-project/issues/97005 (Clang
crashed in ASTContext::getCommonSugaredType).

efriedma-quic commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14171,6 +14171,15 @@ static QualType getCommonSugarTypeNode(ASTContext &Ctx, const Type *X,
+      // qualifiers, both the array and the element type are so-qualified.
+      //
+      // To simplify comparison remove the redundant qualifiers from the array.
+      T.Quals.removeCVRQualifiers(Qualifiers::Const | Qualifiers::Volatile);
```
Tossing the const qualifier is fine if the array element type is already const-qualified... but do we actually know it is at this point?  Or do we need something like `T.Ty = getAsArrayType(getQualifiedType(T))` to make sure the qualifiers get preserved?

spavloff replied:
In the updated version, only qualifiers found in element type are removed.

mizvekov commented:
Regarding the double const , I think this is not a type printer bug per se.

By indiscriminately adding a qualifier to a type, we can produce a type which cannot be written, such as this const qualified array case. As such, we can never do a good job printing these anyway.

There are two issues:
* A single step desugaring with qualifier preservation could cause this. This could be produced as part of the type 'aka' in diagnostics. We probably should add a special case for arrays when single step desugaring.
* When adding const for `constexpr` declarations, we could probably add some new kind of type sugar node which represents how this qualifier was added.

mizvekov commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14171,6 +14171,15 @@ static QualType getCommonSugarTypeNode(ASTContext &Ctx, const Type *X,
 static auto unwrapSugar(SplitQualType &T, Qualifiers &QTotal) {
   SmallVector<SplitQualType, 8> R;
   while (true) {
+    if (const auto *ATy = dyn_cast<ArrayType>(T.Ty)) {
+      // C++ 9.3.3.4p3: Any type of the form "cv-qualifier-seq array of N U" is
+      // adjusted to "array of N cv-qualifier-seq U".
+      // C23 6.7.3p10: If the specification of an array type includes any type
+      // qualifiers, both the array and the element type are so-qualified.
+      //
+      // To simplify comparison remove the redundant qualifiers from the array.
+      T.Quals.removeCVRQualifiers(Qualifiers::Const | Qualifiers::Volatile);
```
This is at the wrong level to fix the problem.
We have the logic for dealing with Array qualifiers in `getCommonArrayElementType`.

That's where we adjust the outer qualifiers for Array types accordingly to the qualifiers of the element type.

Looks like there is a bug in the qualifier algebra there:
```C++
  QX += EX.getQualifiers() - RQ;
  QY += EY.getQualifiers() - RQ;
```
We want to break down those two lines to:
```C++
  QX += EX.getQualifiers();
  QY += EY.getQualifiers();
  QX -= RQ;
  QY -= RQ;
```

I believe this should fix your problem.

spavloff replied:
This change makes sense, it can fix te case of "array of arry", but it cannot fix the case from https://github.com/llvm/llvm-project/issues/97005:
```
bool a;
constexpr const unsigned char c[] = { 5 };
constexpr const unsigned char d[1] = { 0 };
auto b = (a ? d : c);
```
Here types of `d` and `d` do not contain sugar, corresponding canonical types are identical, `getCommonNonSugarTypeNode` is not called and `getCommonArrayElementType` is not called. However their types are not equal and `assert(QX == QY)` fails.

mizvekov replied:
Alright, this needs more thought, but regardless, getCommonArrayElementType is the correct place to fix the problem.

We don't want to add special cases for ArrayTypes in unwrapSugar, because it's in the fast path, and most type nodes are not ArrayTypes, and most types don't even have ArrayTypes in them.

So this is why things are layered this way.

mizvekov commented on clang/test/SemaCXX/constexpr-implicit-const-97005.cpp:
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -triple x86_64-linux-gnu %s
+
+bool a;
+constexpr const unsigned char c[] = { 5 };
+constexpr const unsigned char d[1] = { 0 };
+auto b = (a ? d : c);
+
+constexpr const unsigned char c1[][1] = {{ 5 }};
+constexpr const unsigned char d1[1][1] = {{ 0 }};
+auto b1 = (a ? d1 : c1);
```
Please move this test to `clang/test/SemaCXX/sugar-common-types.cpp`, adjusting according to the function and style of that file.

Ie we want to make sure the inner sugar is preserved in the output.

spavloff replied:
The test has been moved.

spavloff committed:
Address review comments

- Now only qualifiers that are found in array element are cleared,
- Test are moved into a new location.

mizvekov commented:
I posted an alternative fix for this here: https://github.com/llvm/llvm-project/pull/132559

spavloff commented:
Fixed in https://github.com/llvm/llvm-project/pull/132559.
